### PR TITLE
Cleaner test.dart output.

### PR DESCRIPTION
**Pull request by Hixie**
To make running the tests locally more tractable, this adjusts the output of test.dart so that output is now split into sections delimited by `printProgress`, and within each section, unless a test fails (`foundError`) or the section takes a long time (indicating a possible hang), the output is suppressed except for the current line (by repeatedly overprinting the last line of log output).

ANSI color sequences are handled correctly by the code that trims the output to fit on one line.

I've added `--verbose` and `--abort-on-error` to `test.dart`; the former just disables all the fancy stuff I described above; the latter causes `foundError` to be fatal.

In addition, the code is simplified in various ways:

* `verifyNoMissingLicense` in `analyze.dart` doesn't need to count errors any more since each error is reported as it happens.

* some previously bespoke error handling code is routed via common routines in `utils.dart` to reduce code redundancy.

* the "FlutterCompactFormatter" is removed since we weren't using it (we've had it for three years, so if we were going to use it we would have already; as far as I can tell, it had bit-rotted anyway and was no longer functional).

* the code for tracking command output is simplified (it assumed UTF-8 anyway before, so jumping through hoops to keep the data as lists of bytes was pointless).

* added documentation to the top of test.dart.

* simplified how we handle arguments in test.dart.

* fixed some logic errors I happened to notice.

* made some of the function names more consistent.

### CODE REVIEW

**Comment by Hixie** [comment]
@chunhtai any chance I can convince you to review this?

File: dev/bots/utils.dart [hunk 1]
```diff
@@ -2,83 +2,194 @@
...
+  final int seconds = duration.inSeconds - minutes * 60;
+  final int milliseconds = duration.inMilliseconds - (seconds * 1000 + minutes * 60 * 1000);
+  result += '$seconds.${milliseconds.toString().padLeft(3, "0")}s';
+  return result;
+}
+
+typedef PrintCallback = void Function(Object? line);
+typedef VoidCallback = void Function();
 
 // Allow print() to be overridden, for tests.
```
Comment (chunhtai) [inline dev/bots/utils.dart:58]: Can you update the doc?

File: dev/bots/utils.dart [hunk 2]
```diff
@@ -2,83 +2,194 @@
...
-import 'dart:core' as core_internals show print;
+import 'dart:async';
 import 'dart:core' hide print;
 import 'dart:io' as system show exit;
 import 'dart:io' hide exit;
+import 'dart:math' as math;
 
 import 'package:meta/meta.dart';
 
+const Duration _quietTimeout = Duration(minutes: 10); // how long we should hide output between calls to printProgress before just being verbose
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply dev/bots/utils.dart:13]: Here and everywhere

**Review by chunhtai — COMMENTED** [review]
The logs look a lot cleaner than before, thanks for the clean up! Most of my comments are nitpicks. Otherwise they looks good.

File: dev/bots/utils.dart [hunk 3]
```diff
@@ -2,83 +2,194 @@
...
   assert(messages.isNotEmpty);
-  print(redLine);
-  messages.forEach(print);
-  print(redLine);
+  // Make the error message easy to notice in the logs by
+  // wrapping it in a red box.
+  final int width = math.max(15, (hasColor ? stdout.terminalColumns : 80) - 1);
+  print('$red╔═╡${bold}ERROR$reset$red╞═${"═" * (width - 9)}');
+  for (final String message in messages.expand((String line) => line.split('\n'))) {
+    print('$red║$reset $message');
```
[the remark this answers is not in the document]
Comment (Hixie) [reply dev/bots/utils.dart:80]: Yeah, the problem is that the lines are of unbounded length and wrapping them might make the message misleading, but on the other hand just having a line above and below leads to a lot of weird lines so this was the compromise I came up with. \*shrug\*

File: dev/bots/utils.dart [hunk 4]
```diff
@@ -2,83 +2,194 @@
...
-         '▐$reset';
+void printProgress(String message) {
+  _pendingLog.clear();
+  _hideTimer?.cancel();
+  _hideTimer = null;
+  print('$clock $message $reset');
+  if (hasColor) {
+    // We set up a timer to switch to verbose mode when the tests take too long,
+    // so that if a test hangs we can see the logs.
+    // (This is only supported with a true color terminal.)
```
[the remark this answers is not in the document]
Comment (Hixie) [reply dev/bots/utils.dart:123]: I'll make the comment clearer, but the thinking here is that without a color terminal we don't bother doing any of the pretty logic, we just dump everything to the console. (The main case for this is CI logs.)

File: dev/bots/utils.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (Hixie) [reply dev/bots/utils.dart:13]: there are just comments, not dartdocs, so the "we" thing doesn't apply, unless i missed something?

File: dev/bots/utils.dart [hunk 5]
```diff
@@ -2,83 +2,194 @@
...
+// called by foundError and used to implement --abort-on-error in test.dart
+VoidCallback? onError;
 
 bool get hasError => _hasError;
 bool _hasError = false;
 
 Iterable<String> get errorMessages => _errorMessages;
 List<String> _errorMessages = <String>[];
 
+final List<String> _pendingLog = <String>[];
```
[the remark this answers is not in the document]
Comment (Hixie) [reply dev/bots/utils.dart:70]: fixed

File: dev/bots/utils.dart [hunk 6]
```diff
@@ -2,83 +2,194 @@
...
+  // overwriting it. To do this in color mode we have to carefully measure the
+  // line of text ignoring color codes, which is what the parser below does.
+  if (_hideTimer != null) {
+    _pendingLog.add(message.toString());
+    String line = '$message'.trimRight();
+    final int start = line.lastIndexOf(_lineBreak) + 1;
+    int index = start;
+    int length = 0;
+    while (index < line.length && length < stdout.terminalColumns) {
+      if (line.codeUnitAt(index) == 0x1B) { // ESC
```
[the remark this answers is not in the document]
Comment (Hixie) [reply dev/bots/utils.dart:148]: Done.

File: dev/bots/utils.dart [hunk 7]
```diff
@@ -2,83 +2,194 @@
...
+    stdout.writeln(message);
+  }
 }
 
+// THE FOLLOWING CODE IS A VIOLATION OF OUR STYLE GUIDE
+// BECAUSE IT INTRODUCES A VERY FLAKY RACE CONDITION
+// https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#never-check-if-a-port-is-available-before-using-it-never-add-timeouts-and-other-race-conditions
+// DO NOT USE THE FOLLOWING FUNCTIONS
+// DO NOT WRITE CODE LIKE THE FOLLOWING FUNCTIONS
+// TODO(ianh): determine why this code was added and find a better solution
```
[the remark this answers is not in the document]
Comment (Hixie) [reply dev/bots/utils.dart:191]: done

File: dev/bots/utils.dart (hunk 1 above)
Comment (Hixie) [reply dev/bots/utils.dart:58]: done

**Comment by Hixie** [comment]
PTAL

File: dev/bots/utils.dart [hunk 8]
```diff
@@ -2,87 +2,210 @@
...
+  if (_hideTimer != null) {
+    _pendingLogs.add(message.toString());
+    String line = '$message'.trimRight();
+    final int start = line.lastIndexOf(_lineBreak) + 1;
+    int index = start;
+    int length = 0;
+    while (index < line.length && length < stdout.terminalColumns) {
+      if (line.codeUnitAt(index) == kESC) {
+        index += 1;
+        if (index < line.length && line.codeUnitAt(index) == 0x5B) { // [
```
Comment (chunhtai) [inline dev/bots/utils.dart:162]: Can you also convert this and below to const as well?

**Review by chunhtai — APPROVED** [review]
LGTM

File: dev/bots/utils.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (Hixie) [reply dev/bots/utils.dart:13]: fair enough. i'll do a pass and see if i can make the comments clearer.

File: dev/bots/utils.dart (hunk 8 above)
Comment (Hixie) [reply dev/bots/utils.dart:162]: yup

**Pull request merged**